packages/compass-connections-navigation/src/connections-navigation-tree.tsx

```diff
@@ -391,14 +401,37 @@ const ConnectionsNavigationTree: React.FunctionComponent<
   const listRef = useRef<List | null>(null);
   const id = useId();
 
+  // there are different workspace variants, not all have a connectionId or namespace
+  // we destructure these properties here to not deal with this uncertainty repeatedly
+  const { activeConnectionId, activeNamespace, activeWorkspaceType } =
+    useMemo(() => {
+      const {
+        connectionId: activeConnectionId,
+        namespace: activeNamespace,
+        type: activeWorkspaceType,
+      } = activeWorkspace as {
+        connectionId?: string;
+        namespace?: string;
+        type: string;
+      };
+      return { activeConnectionId, activeNamespace, activeWorkspaceType };
+    }, [activeWorkspace]);
+
   useEffect(() => {
-    if (activeNamespace) {
-      for (const connection of connections) {
-        onDatabaseExpand(connection.connectionInfo.id, activeNamespace, true);
+    if (activeWorkspace) {
+      if (activeConnectionId) {
+        onConnectionExpand(activeConnectionId, true);
+
+        if (activeNamespace) {
+          onDatabaseExpand(activeConnectionId, activeNamespace, true);
+        }
       }
     }
-  }, [activeNamespace, onDatabaseExpand, connections[0]?.connectionInfo.id]);
-  // TODO(COMPASS-7775): the we'll need something similar to expand the active connection
+    // onConnectionExpand is excluded from the deps intentionally -
+    // we only want to expand as a reaction to workspace change, not to a collapse action
+    // otherwise active connections are re-expanded
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
Would it then make sense to wrap onConnectionExpand in a useRef?

Reply:
only for this part - the autoexpanding. but the function is also used for intentional collapse/expand

Reply:
hmm. yeah, we could do it. the intentional collapse/expand most likely doesn't have another effect depending on it, and it could always use the current. let me try it.

Reply:
works like a charm! thank you @himanshusinghs !